Webgpu support

dc-dc-dc opened the pull request:
In collab with @[martinloretzzz](https://github.com/geohot/tinygrad/commits?author=martinloretzzz)

Closes #1030

tinyb0t commented:
Changes made in `tinygrad/`:
```
------------------------------------------------------------
files                             insertions       deletions
------------------------------------------------------------
tinygrad/codegen/linearizer.py             3               3
tinygrad/jit.py                            1               1
tinygrad/runtime/ops_webgpu.py           133               0
------------------------------------------------------------
total                                    137               4
------------------------------------------------------------
lines added in the tinygrad folder: 133
```

jumbojets commented on tinygrad/runtime/ops_webgpu.py:
```diff
@@ -0,0 +1,133 @@
+
+class RawWebGPUBuffer(RawBufferCopyIn):
+  def __init__(self, size:int, dtype:DType):
+    assert dtype not in [dtypes.int8,dtypes.uint8,dtypes.int64,dtypes.uint64,dtypes.float64], f"dtype {dtype} not supported on WEBGPU"
```
could you use arbitrary float https://www.w3.org/TR/WGSL/#example-bbb80169 for float64?

jumbojets commented:
Could float64 and float4 could both supported via [arbitrary float](https://www.w3.org/TR/WGSL/#example-bbb80169) and [`vec4f`](https://www.w3.org/TR/WGSL/#vec4-builtin) respectively?

jumbojets commented:
Could float64 and float4 could both supported via [arbitrary float](https://www.w3.org/TR/WGSL/#example-bbb80169) and [`vec4f`](https://www.w3.org/TR/WGSL/#vec4-builtin) respectively?

geohot commented:
Woah didn't realize this was ready. Let's see

geohot commented:
Is there a reason this can't use C style codegen and needs a new one?

geohot commented:
This is honestly an amazing PR. Raised the bounty after the fact to $500, I think this can deliver more value than I first thought.

dc-dc-dc commented:
- refactored to utilize CStyleCodegen, added some extra logic and moved some stuff around
- I created a puppeteer script that loads up the page and tests the output, but looks like the CI version of chrome is not properly creating a gpu device so will just keep the compilation and leave the puppeteer script for whenever they add support for it in CI.  
CI seems like its still having issues so I posted an issue to the `wgpu-py` repo to see if they have some remedies to help resolve

dc-dc-dc commented:
switched to macos-13 and disable the conv tests

dc-dc-dc commented:
forcing wgpu to use metal resolved some issues in CI, just had to disable the conv2d tests but everything else is running fine

geohot commented:
Yea figured it might be annoying, just the compile test is fine for now until someone makes this work.

geohot commented:
Congrats! Will merge when tests pass.

Either post PayPal e-mail here or e-mail george@tinygrad.org to claim and state what split you want.
